src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,261 @@
+  combineLatest as observableCombineLatest,
+  EMPTY,
+  Observable,
+  of as observableOf,
+} from "rxjs";
+import {
+  catchError,
+  mergeMap,
+  tap,
+} from "rxjs/operators";
+import Manifest from "../../../manifest";
+import {
+  disposeSourceBuffer,
+  getSourceBuffer$,
+} from "./get_source_buffer";
+import log from "./log";
+import {
+  loadInitThumbnail,
+  loadThumbnails
+} from "./segment_utils";
+import {
+  getThumbnailTrack,
+  getWantedThumbnails,
+  IThumbnail,
+  IThumbnailTrack,
+} from "./thumbnail_track_utils";
+
+const PPromise = typeof Promise === "function" ? Promise :
+                                                 pinkie;
+
+interface IJob { thumbnailTrack: IThumbnailTrack;
+                 thumbnails: IThumbnail[];
+                 stop: () => void; }
+
+// Returned error when rejecting
+class VideoThumbnailLoaderError extends Error {
+  public readonly name : "VideoThumbnailLoaderError";
+  public readonly message : string;
+  public readonly code : string;
+
+  /**
+   * @param {string} code
+   * @param {string} reason
+   * @param {Boolean} fatal
+   */
+  constructor(code : string, message : string) {
+    super();
+    Object.setPrototypeOf(this, VideoThumbnailLoaderError.prototype);
+    this.name = "VideoThumbnailLoaderError";
+    this.code = code;
+    this.message = message;
+  }
+}
+
+/**
+ * This tool, as a supplement to the RxPlayer, intent to help creating thumbnails
+ * from a video source.
+ *
+ * The tools will extract a "thumbnail track" either from a video track (whose light
+ * chunks are adapted from such use case) or direclty from the media content.
+ */
+export default class VideoThumbnailLoader {
+  private readonly _thumbnailVideoElement: HTMLVideoElement;
+
+  private _manifest: Manifest;
+  private _currentJob?: IJob;
+
+  constructor(videoElement: HTMLVideoElement,
+              manifest: Manifest) {
+    this._thumbnailVideoElement = videoElement;
+    this._manifest = manifest;
+  }
+
+  /**
+   * Set time of thumbnail video media element :
+   * - Remove buffer when too much buffered data
+   * - Search for thumbnail track element to display
+   * - Load data
+   * - Append data
+   * Resolves when time is set.
+   * @param {number} time
+   * @returns {Promise}
+   */
+  setTime(time: number): Promise<unknown> {
+    return new PPromise((done, failed) => {
+      this._setTime({ time, done, failed }, this._thumbnailVideoElement);
+    });
+  }
+
+  /**
+   * Dispose thumbnail loader.
+   * @returns {void}
+   */
+  dispose(): void {
+    disposeSourceBuffer();
+    return this._currentJob?.stop();
+  }
+
+  private _setTime = (payload: { time: number;
+                                 done: (time: number) => void;
+                                 failed: (err: unknown) => void; },
+                      videoElement: HTMLVideoElement
+  ): void => {
+    const { time, done, failed } = payload;
+    if (time === this._thumbnailVideoElement.currentTime) {
+      return done(time);
```

Comment:
What about the initial time (`0`)?


I think we should still continue even if we are already at the right time. Worst case scenario, we will see that we already downloaded the corresponding segment.

Reply:
Done